sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala

```diff
@@ -171,15 +171,21 @@ private[csv] class CsvOutputWriter(
 
   private var univocityGenerator: Option[UnivocityGenerator] = None
 
-  override def write(row: InternalRow): Unit = {
-    val gen = univocityGenerator.getOrElse {
-      val charset = Charset.forName(params.charset)
-      val os = CodecStreams.createOutputStreamWriter(context, new Path(path), charset)
-      val newGen = new UnivocityGenerator(dataSchema, os, params)
-      univocityGenerator = Some(newGen)
-      newGen
-    }
+  if (params.headerFlag) {
+    val gen = getGen()
+    gen.writeHeaders()
+  }
 
+  private def getGen(): UnivocityGenerator = univocityGenerator.getOrElse {
+    val charset = Charset.forName(params.charset)
+    val os = CodecStreams.createOutputStreamWriter(context, new Path(path), charset)
+    val newGen = new UnivocityGenerator(dataSchema, os, params)
+    univocityGenerator = Some(newGen)
+    newGen
+  }
+
+  override def write(row: InternalRow): Unit = {
+    val gen = getGen()
```

Comment:
Wait .. is this going to create `UnivocityGenerator` for each record?

Reply:
Ah, it's `getOrElse`. Okay but still can we simplify this logic? Looks a bit confusing. For instance, I think we can do this with lazy val.

Reply:
@HyukjinKwon Do you mean creating generator in lazy val?
```
lazy val univocityGenerator = {
    val charset = Charset.forName(params.charset)
    val os = CodecStreams.createOutputStreamWriter(context, new Path(path), charset)
    new UnivocityGenerator(dataSchema, os, params)
}
```
The problem is in the close method, you will have to call `univocityGenerator.close()` in the method. If the lazy val wasn't instantiated before (empty partition and the `header` option is `false`), the generator will be created and closed immediately. And as a result, you will get an empty file for the empty partition. That's why I prefer the approach with `Option[UnivocityGenerator]` in https://github.com/apache/spark/pull/23052.

Reply:
I see the problem. OrcFileFormat uses a flag approach. For instance:



```
private var isGeneratorInitiated = false

lazy val univocityGenerator = {
    val charset = Charset.forName(params.charset)
    val os = CodecStreams.createOutputStreamWriter(context, new Path(path), charset)
    new UnivocityGenerator(dataSchema, os, params)
}

if (isGeneratorInitiated) {
    univocityGenerator.close()
}
```

Should be okay to stick to it.

Reply:
Frankly speaking I don't see any reasons for this. For now we have 2 flags actually - `isGeneratorInitiated` and another one inside of lazy val. And 2 slightly different approaches - with the `Option` type in Json and Text, and lazy val + flag in Orc and CSV.

Reply:
Yeah we have two different approaches, both of which are fine IMHO. I think it's reasonable to clean that up in a follow-up if desired. WDYT @HyukjinKwon ?

Reply:
OK. Shouldn't be a big deal.